File: leetcode-submissions/120_Triangle.cpp
class Solution {
public:
vector<vector<int>>dp;
vector<vector<int>>arr;
int f(int row ,int col){
 //base case 
 if(row ==arr.size()-1 ){
     return arr[row][col];
      }
      if(dp[row][col]!=-1) return dp[row][col];
      return dp[row][col]=arr[row][col]+min(f(row+1,col),f(row+1,col+1));
}
    int minimumTotal(vector<vector<int>>& triangle) {
        arr=triangle;
        dp.clear();
        dp.resize(205,vector<int>(205,-1));
        return f(0,0);
    }
};

File: leetcode-submissions/120_Triangle.ts
function minimumTotal(triangle: number[][]): number {
    const n :number = triangle.length;
    const dp :number[][]  = Array.from({length:202}, ()=>Array.from({length:202}, ()=>-1));
    return solve(triangle , 0 , 0 , n, dp);
};
function solve(t:number[][], i :number , j:number , n :number, dp:number[][]):number{
    if(i===n-1){
        return t[i][j];
    }
    if(i>=n){
        return Number.MAX_SAFE_INTEGER;
    }
    if(dp[i][j]!==-1){
        return dp[i][j];
    }
    const firstRule :number = t[i][j]+solve(t, i+1, j, n , dp);
    const secondRule :number = t[i][j]+solve(t, i+1, j+1, n, dp);
    return dp[i][j] = Math.min(firstRule , secondRule);
}